src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public  class ServerSentEventsMessageParser
+    {
+
+        public static ParsePhase ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            const byte ByteCR = (byte)'\r';
+            const byte ByteLF = (byte)'\n';
+            const byte ByteSpace = (byte)' ';
+            bool firstLine = true;
+            byte[] data = null;
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            MessageType messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                var span = reader.Span;
+                var backup = reader;
+                var ch1 = reader.Take();
+                var ch2 = reader.Take();
+                if (ch1 == ByteCR)
+                {
+                    if (ch2 == ByteLF)
+                    {
+                        consumed = buffer.End;
+                        message = new Message(data, messageType);
+                        return ParsePhase.Completed;
+                    }
+
+                    //When we are only missing the final \n
+                    return ParsePhase.Incomplete;
+                }
+
+                reader = backup;
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    // Not there
+                    return ParsePhase.Incomplete;
+                }
+
+                // Make sure LF is included in lineEnd
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                if (firstLine)
+                {
```

Comment:
Need to check `data:` prefix here as well. Again, there's a clever way we could do the whole thing (checking `data:` **and** figuring out what the type is) if you've got some cycles after getting the simple method implemented and want to try something ;).